### ISSUE TITLE
Extremely slow `repo:` search on Sourcegraph.com

### ISSUE THREAD

**Issue by rvantonder** [sourcegraph/sourcegraph#34946]
It takes 15 seconds to show results for `repo:sourcegraph/sourcegraph` [link](https://sourcegraph.com/search?q=context:global+repo:sourcegraph/sourcegraph&patternType=literal) but results are instant for `repo:github.com/sourcegraph/sourcegraph` [link](https://sourcegraph.com/search?q=context:global+repo:github.com/sourcegraph/sourcegraph&patternType=literal). This seems so slow that it seems like a regression/bug. Labeling as bug until we learn otherwise.

https://user-images.githubusercontent.com/888624/166839793-459fdf61-8164-4204-9953-3cd68a3e4e21.mp4

**Comment by sourcegraph-bot-2** [comment]
Heads up @benvenker @lguychard - the "team/search-product" label was applied to this issue.

**Comment by sourcegraph-bot-2** [comment]
Heads up @jjeffwarner - the "team/search-core" label was applied to this issue.

**Comment by keegancsmith** [comment]
TL;DR: postgres is picking a poor query plan

This is indeed a regression. It's a little hard to debug, since we are trying out datadog and the traces have no logs. But luckily they have tags so I can try doing an explain on the query. I experimented by running simplified queries:

This query without order by was instant:

```sql
SELECT repo.id,repo.name,repo.private,repo.stars
FROM repo
WHERE repo.deleted_at IS NULL AND repo.blocked IS NULL AND ((lower(name) LIKE '%sourcegraph/sourcegraph%'));
```

With order by was super slow:

```sql
SELECT repo.id,repo.name,repo.private,repo.stars
FROM repo
WHERE
  repo.deleted_at IS NULL AND repo.blocked IS NULL AND ((lower(name) LIKE '%sourcegraph/sourcegraph%')
 AND  NOT archived)
ORDER BY stars DESC NULLS LAST ,  id DESC LIMIT 500 OFFSET 0
```

Here are the respective explain outputs:

```
Bitmap Heap Scan on repo  (cost=753.89..63619.55 rows=58642 width=53) (actual time=29.429..33.208 rows=55 loops=1)
  Recheck Cond: ((lower((name)::text) ~~ '%sourcegraph/sourcegraph%'::text) AND (deleted_at IS NULL) AND (blocked IS NULL))   Rows Removed by Index Recheck: 1104
  Filter: (NOT archived)
  Rows Removed by Filter: 27
  Heap Blocks: exact=1133
  ->  Bitmap Index Scan on erik_test_remove7  (cost=0.00..739.23 rows=59806 width=0) (actual time=28.707..28.708 rows=1814 loops=1)
        Index Cond: (lower((name)::text) ~~ '%sourcegraph/sourcegraph%'::text)
Planning Time: 0.346 ms
Execution Time: 33.381 ms
```

```
Limit  (cost=0.43..9001.52 rows=500 width=53) (actual time=12.680..14925.569 rows=55 loops=1)
  ->  Index Scan using repo_stars_desc_id_desc_idx on repo  (cost=0.43..1055684.39 rows=58642 width=53) (actual time=12.679..14925.509 rows=55 loops=1)
        Filter: ((NOT archived) AND (lower((name)::text) ~~ '%sourcegraph/sourcegraph%'::text))
        Rows Removed by Filter: 5881518
Planning Time: 0.490 ms
Execution Time: 14925.654 ms
```

cc @tsenart our export at looking at slow postgres queries against the repo table as well as the pagination king.

Also cc @eseliger since a temporary index you introduced is being used by the simple query I added.

**Comment by keegancsmith** [comment]
The `vacuum analyze repo` fixed the problem, it is instant now.

@eugeniaft occasionally running the vacuum would likely help perf for search at could fit in as part of our team's charter.

**Comment by keegancsmith** [comment]
For posterity here is the query plan now

```
Limit  (cost=776.56..777.81 rows=500 width=53) (actual time=24.880..24.890 rows=55 loops=1)
   ->  Sort  (cost=776.56..777.98 rows=568 width=53) (actual time=24.879..24.883 rows=55 loops=1)
         Sort Key: stars DESC NULLS LAST, id DESC
         Sort Method: quicksort  Memory: 32kB
         ->  Bitmap Heap Scan on repo  (cost=105.41..750.57 rows=568 width=53) (actual time=21.998..24.827 rows=55 loops=1)
               Recheck Cond: ((lower((name)::text) ~~ '%sourcegraph/sourcegraph%'::text) AND (deleted_at IS NULL) AND (blocked IS NULL))
               Rows Removed by Index Recheck: 1104
               Filter: (NOT archived)
               Rows Removed by Filter: 27
               Heap Blocks: exact=1133
               ->  Bitmap Index Scan on erik_test_remove7  (cost=0.00..105.26 rows=580 width=0) (actual time=21.599..21.600 rows=1186 loops=1)
                     Index Cond: (lower((name)::text) ~~ '%sourcegraph/sourcegraph%'::text)
 Planning Time: 3.539 ms
 Execution Time: 25.041 ms
```

@eseliger it seems we are using your `erik_test_remove7` index. Should we make that index official and ship it in our code?

**Comment by eseliger** [comment]
yes! sorry there's a few of these that I still need to cofidy 🙃 Will try to do that soon!
